### PR TITLE
Add a readme for k8s autodiscover provider

### PR DESCRIPTION
**Pull request by MichaelKatsoulis**
## What does this PR do?

This PR creates a markdown file explaining the high and low level functionality of Kubernetes autodiscover provider.
It consists of three major segments.

1. Templates Based Autodiscover
2. Hints Based Autodiscover
3. Autodiscover provider in Elastic Agent

## Why is it important?

The intention is to have a technical document which will help everyone who wants to understand how Kubernetes autodiscover works.

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have made corresponding change to the default configuration files
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

### CODE REVIEW

File: libbeat/autodiscover/README.md [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+3. Autodiscover provider in Elastic Agent
+
+## Templates Based Autodiscover
+
+As the name suggests, user needs to set a template to indicate to autodiscover provider what to do.
+There is one configuration variable that differentiates in a way how the autosicover process is performed.
+This variable is `unique`
+
+### Autodiscover with LeaderElection
+When setting `unique: true` the Leader Election mechanism is activated. That way **only** the beat process that will gain the leader lease will enable the provided template.
```
Comment (ChrsMark) [inline libbeat/autodiscover/README.md:31]:
```suggestion
When setting `unique: true` the Leader Election mechanism is activated. That way **only** the Beat instance that will gain the leader lease/lock will enable the provided template.
```

File: libbeat/autodiscover/README.md [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+```
+
+##### How it works
+
+We will deep dive in the internals of [libbeat kubernetes autodiscover provider](https://github.com/elastic/beats/tree/master/libbeat/autodiscover/providers/kubernetes).
+
+We will use metricbeat as an example.
+
+Step-by-step walkthrough
+1. Kubernetes provider `init` function [adds](https://github.com/elastic/beats/blob/master/libbeat/autodiscover/providers/kubernetes/kubernetes.go#L46) the provider in the autodiscover providers registry at startup. For Kubernetes provider a `AutodiscoverBuilder` func is passed as an argument.
```
Comment (ChrsMark) [inline libbeat/autodiscover/README.md:61]:
```suggestion
1. Kubernetes provider `init` function [adds](https://github.com/elastic/beats/blob/master/libbeat/autodiscover/providers/kubernetes/kubernetes.go#L46) the provider in the autodiscover providers registry at startup. For Kubernetes provider an `AutodiscoverBuilder` func is passed as an argument.
```

File: libbeat/autodiscover/README.md [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+2. Metricbeat calls `NewAutodiscover` [function](https://github.com/elastic/beats/blob/master/metricbeat/beater/metricbeat.go#L183) which checks in the config for enabled providers and [builds](https://github.com/elastic/beats/blob/master/libbeat/autodiscover/provider.go#L90) them on by one by calling the `AutodiscoverBuilder` func.
```
Comment (ChrsMark) [inline libbeat/autodiscover/README.md:62]:
```suggestion
2. Metricbeat calls `NewAutodiscover` [function](https://github.com/elastic/beats/blob/master/metricbeat/beater/metricbeat.go#L183) which checks in the config for enabled providers and [builds](https://github.com/elastic/beats/blob/master/libbeat/autodiscover/provider.go#L90) them one by one, calling the `AutodiscoverBuilder` func.
```

File: libbeat/autodiscover/README.md [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+3. Kubernetes `AutodiscoverBuilder` creates and returns a [Kubernetes Provider struct](https://github.com/elastic/beats/blob/master/libbeat/autodiscover/providers/kubernetes/kubernetes.go#L131) which is then added in a Autodiscover manager struct.
```
Comment (ChrsMark) [inline libbeat/autodiscover/README.md:63]:
```suggestion
3. Kubernetes `AutodiscoverBuilder` creates and returns a [Kubernetes Provider struct](https://github.com/elastic/beats/blob/master/libbeat/autodiscover/providers/kubernetes/kubernetes.go#L131) which is then added to an Autodiscover manager struct.
```

File: libbeat/autodiscover/README.md [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+4. When unique is set to true [NewLeaderElectionManager](https://github.com/elastic/beats/blob/master/libbeat/autodiscover/providers/kubernetes/kubernetes.go#L141) is set as the eventManager of Kubernetes Provider.
+4. Metricbeat [starts](https://github.com/elastic/beats/blob/master/metricbeat/beater/metricbeat.go#L249) the Autodiscover manager which starts for Kubernets provider the [leaderElectionManager](https://github.com/elastic/beats/blob/master/libbeat/autodiscover/providers/kubernetes/kubernetes.go#L326). Before starting the providers it also starts a worker for listening of events that will be published by the eventers of each provider.
+5. OnStartedLeading is executed when the specific metricbeat instance gains the leader election lock. [StartLeading](startLeading) cretaes a bus event with `"start":    true,`  and publishes it. The template configurations is also added in this event.
```
Comment (ChrsMark) [inline libbeat/autodiscover/README.md:66]:
```suggestion
5. `OnStartedLeading` is executed when the specific metricbeat instance gains the leader election lock. [StartLeading](startLeading) cretaes a bus event with `"start":    true,`  and publishes it. The template configurations is also added in this event.
```

File: libbeat/autodiscover/README.md [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+7. For each config the worker checks if it already [exists](https://github.com/elastic/beats/blob/master/libbeat/autodiscover/autodiscover.go#L225)(It was already handled). If at least of the configs of the event does not then the config is marked as `updated`.
+8. The runners list get [reloaded](https://github.com/elastic/beats/blob/4b1f69923b3f2abbbf1860295fe5dbff7db3d63c/libbeat/cfgfile/list.go#L54). It is checked from the list of current runners if each config is handled by one of them.
+   If no runner is handling that config a new runner will [start](https://github.com/elastic/beats/blob/4b1f69923b3f2abbbf1860295fe5dbff7db3d63c/libbeat/cfgfile/list.go#L107). If some runners are no longer needed will be removed.
+9. Starting of a runner starts under the hood the metricbeat [module](https://github.com/elastic/beats/blob/4b1f69923b3f2abbbf1860295fe5dbff7db3d63c/metricbeat/mb/module/runner.go#L76) with the specific metricsets as they are set in the template.
```
Comment (ChrsMark) [inline libbeat/autodiscover/README.md:71]: Same happens for filebeat and hearbeat and not only for metricbeat. Maybe you can somehow make it more generic here?

File: libbeat/autodiscover/README.md [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+
+### Autodiscover without LeaderElection
+
+When setting `unique: false` or not setting it at all the Leader Election mechanism is disabled. That way all the beat instances will enable the provided template. Or at least the ones that match a condition.
```
Comment (ChrsMark) [inline libbeat/autodiscover/README.md:75]:
```suggestion
When setting `unique: false` or not setting it at all the Leader Election mechanism is disabled. That way all Beat instances will enable the provided template. Or at least the ones that match a condition.
```

File: libbeat/autodiscover/README.md [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+* [OnDelete](https://github.com/elastic/beats/blob/4b1f69923b3f2abbbf1860295fe5dbff7db3d63c/libbeat/autodiscover/providers/kubernetes/pod.go#L160) is called when there is a pod deletion and it ecexutes emit with `stop` flag. Emit adds the pod to a new event with ``stop: true`` and publishes it. Step 7 follows.
+* [OnUpdate](https://github.com/elastic/beats/blob/4b1f69923b3f2abbbf1860295fe5dbff7db3d63c/libbeat/autodiscover/providers/kubernetes/pod.go#L146) is called when there is update in the an existing resource and it ecexutes at first emit with `stop` flag to generate a stop event and then with `start` flag to generate a start event.
+
+
+## Hints Based Autodiscover
+
+https://www.elastic.co/guide/en/beats/metricbeat/master/configuration-autodiscover-hints.html
+
+Metricbeat and Filebeat support autodiscover based on hints from the provider. The hints system looks for hints in Kubernetes Pod annotations which have the prefix co.elastic.metrics. As soon as the container starts, Metricbeat/Filebeat will check if it contains any hints and launch the proper config for it. Hints tell Metricbeat/Filebeat how to get metrics for the given container.
```
Comment (ChrsMark) [inline libbeat/autodiscover/README.md:124]: This also applies for Heartbeat.

File: libbeat/autodiscover/README.md [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+
+Template based autodiscover of Kubernetes resources is only supported in standalone mode as of now.
+It is not part of the Kubernetes Integration yet.
+
+Hints based autodiscover is not supported yet.
+
+### Template based autodiscover
+
+Example:
+As en example we will use gain redis module.
```
Comment (ChrsMark) [inline libbeat/autodiscover/README.md:205]:
```suggestion
As an example we will use gain redis module.
```

Comment (ChrsMark) [inline libbeat/autodiscover/README.md:205]:
```suggestion
As en example we will use again redis module.
```

Comment (ChrsMark) [inline libbeat/autodiscover/README.md:205]:
```suggestion
As en example we will use gain the `redis` module.
```

File: libbeat/autodiscover/README.md [hunk 10]
```diff
@@ -0,0 +1,260 @@
...
+In agent.yml(configmap) an extra input block needs to be added.
+```
+      # Add extra input blocks here, based on conditions
```
Comment (ChrsMark) [inline libbeat/autodiscover/README.md:208]: indentation?

File: libbeat/autodiscover/README.md [hunk 11]
```diff
@@ -0,0 +1,260 @@
...
+10. GenerateHints[https://github.com/elastic/beats/blob/eff92354db783001880f4bade9f59942fca747ba/libbeat/autodiscover/builder/helper.go#L213] function looks into the event's annotations. A [hints map](https://github.com/elastic/beats/blob/eff92354db783001880f4bade9f59942fca747ba/libbeat/autodiscover/builder/helper.go#L226) is created with all hints and returned.
+11. From those hints, configs are [created](https://github.com/elastic/beats/blob/master/libbeat/autodiscover/builder.go#L97) in the same form as in `Autodiscover without LeaderElection` step 8.
+    They contain the same information as if they where set explicitly in the metricbeat configureation but actually derive from the pod annotations.
+12. Those configs are then [added](https://github.com/elastic/beats/blob/master/libbeat/autodiscover/providers/kubernetes/kubernetes.go#L197) in the event and gets published.
+13. The process after that is same as in `Autodiscover without LeaderElection` step 9 and onward.
+
+
+
+##  Autodiscover provider in Elastic Agent
```
Comment (ChrsMark) [inline libbeat/autodiscover/README.md:190]: I would suggest moving this part under `elastic-agent` codebase so as to be close to the provider's code. Then you can cross-link it from libbeat's docs.

**Review by ChrsMark — COMMENTED** [review]
Very nice that we finally have tech documentation for this codebase!
I left some suggestions about wording. Also moving Agent's part to the respective codebase directory is something that seems more suitable.

Also some comments:
1. leader_election is described as part of libbeat's provider but we have also a dedicated provider for Agent. Maybe it is worth documenting it to since you are on it.
2. In addition to this, there is also the `kubernetes_secrets` provider for Agent and `kubernetes_secrtes` backend for Beats, which is something we could document too as part of this effort.

**Comment by MichaelKatsoulis** [comment]
@ChrsMark @tetianakravchenko I updated all links of code in the READMEs with permalinks. I think it is safe, as at least the text will always correspond to the code. 
Maybe master at some point will diverge a bit as we update the code.
I don't consider it to be a big issue. If we add a new feature or if we completely refactor a function/process then we should update the readmes as well.

**Review by tetianakravchenko — APPROVED** [review]

**Pull request merged**